src/group_theory/nilpotent.lean

```diff
@@ -675,3 +685,53 @@ begin
   rw [eq_bot_iff, ←hn],
   exact derived_le_lower_central n,
 end
+
+section classical
+
+open_locale classical
+
+/-- A custom induction principle fintypes. The base case is a subsingleton type,
+and the induction step is for non-trivial types, and one can assume the hypothesis for
+smaller groups (via `fintype.card`).
+-/
+@[elab_as_eliminator]
+lemma fintype.induction_subsingleton_or_nontrivial
+  {P : Π α [fintype α], Prop}
+  (α : Type*)
+  [fintype α]
+  (hbase : ∀ α [fintype α] [subsingleton α], by exactI P α)
+  (hstep : ∀ α [fintype α] [nontrivial α],
+    by exactI ∀ (ih : ∀ β [fintype β], by exactI ∀ (h : fintype.card β < fintype.card α), P β),
+    P α) :
+  P α :=
+begin
+  obtain ⟨ n, hn ⟩ : ∃ n, fintype.card α = n := ⟨fintype.card α, rfl⟩,
+  unfreezingI { induction n using nat.strong_induction_on with n ih generalizing α },
+  casesI (subsingleton_or_nontrivial α) with hsing hnontriv,
+  { apply hbase, },
+  { apply hstep,
+    introsI β _ hlt,
+    rw hn at hlt,
+    exact (ih (fintype.card β) hlt _ rfl), }
+end
+
+/-- A p-group is nilpotent -/
+lemma is_nilpotent.of_p_group [fintype G] {p : ℕ} (hp : fact (nat.prime p)) (h : is_p_group p G) :
+  is_nilpotent G :=
+begin
+  unfreezingI
+  { revert _inst_1,
+    induction _inst_3 using fintype.induction_subsingleton_or_nontrivial with G _ _ G }; resetI,
```

Comment:
Please don't refer to `_inst_n` explicitly, because these auto-generated names aren't very stable.
Note that you can name instances explicitly in the statement, like `[hG : group G]` etc.

Reply:
Yes, of course. I think I first wanted to hear if the induction principle is useful, and if it is, do some more investigation why this can't just be `induction G`, which would be prettier, and obliterate the need for `revert` before.